[AIRFLOW-3964][AIP-17] Build smart sensor

YingboWang opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW-3964/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3964
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP-17](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-17%3A+Consolidate+and+de-duplicate+sensor+tasks+in+airflow+Smart+Sensor)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

- Add a new mode called “smart sensor mode”. In smart sensor mode, instead of holding a long running process for each sensor and poking periodically, a sensor will only store poke context at sensor_instance table and then exits with a ‘sensing’ state. 
- When the smart sensor mode is enabled, a special set of builtin smart sensor DAGs (named smart_sensor_group_shard_xxx) is created by the system; These DAGs contain SmartSensorOperator task and manage the smart sensor jobs for the airflow cluster. The SmartSensorOperator task can fetch hundreds of ‘sensing’ instances from sensor_instance table and poke on behalf of them in batches. Users don’t need to change their existing DAGs.
- The smart sensor mode currently supports NamedHivePartitionSensor and MetastorePartitionSensor however it can easily be extended to support more sensor classes. 
- Smart sensor mode on/off, the list of smart sensor enabled classes, and the number of SmartSensorOperator tasks can be configured in airflow config.
- Sensor logs in smart sensors are populated to each task instance log UI.

Architecture:
![image](https://user-images.githubusercontent.com/2688818/82953095-09012180-9f5f-11ea-9ded-f472d160af43.png)



This PR include following changes:

1. DB schema:
  a. DB schema change: Add sensor_instance table for persisting sensor poke_context (argument/information needed by poke function) and execution_context (user configured execution context such as timeout, email).
  b. Add SensorInstance class. This class implement function of persisting context mentioned above to airflow metaDB.

2. New operator SmartSensorOperator. This is the key component for executing sensors when an airflow cluster is running in "smart_sensor" mode. This operator take over all poking jobs using sensor_instance table record. It will get context for multiple sensors from DB instead of parsing DAG file, poke, change ti state when task finished either by poking succeeded or timeout or exception, send alert when needed. 

3. Changes in operators:
  a. BaseOperator: Add is_smart_sensor_compatible() with default value as False. It return True if current task is eligible for using smart sensor.
  b. BaseSensorOperator
      i. Add register_in_sensor_service(), it will get sensor context value and call sensor_instance register function.
      ii. Add tuple execution_context_fields to identify general execution context required for handling sensor tasks.
      iii. Add function to get the poke_context and execution_context for sensor jobs
  c. For specific sensor class that need to support smart sensor: (in this PR NamedHivePartitionSensor and MetastorePartitionSensor)
      i. Add poke_context_fields that specifying class attributes that need to be used for poking
      ii. Overwrite is_smart_sensor_compatible()
      (ps: Instruction above can be used to add support to new sensor operator)

4. Add new state to airflow task_instance: Add ‘sensing’ state. Update UI for the new state as well. 'sensing' state implies that 
  a. The sensor poke/execution context has been persisted into DB.
  b. This sensor is no longer holding single process, the slot would be released for tasks in 'sensing' state
  c. Smart sensor will handle all 'sensing' tasks 

5. Change taskinstance.py to add smart sensor workflow:
  a. Add checking after pre_execute for smart sensor compatibility. 
  b. Register qualified sensor jobs(which will set the task instance state to ‘sensing’ state) and exit running tasks without poking it. 
  c. In order to support email_alert in smart sensor, change email_alert to make it callable from both traditional task instance and smart sensor operator

6. Manage smart sensor as a service in an airflow cluster:
  a. Turn smart sensor mode on/off by configuring use_smart_sensor configuration
  b. Use native DAGs with a smart sensor operator to spin up and rotate the smart sensor jobs for the airflow cluster. These DAGs were generated from airflow source code and do not need any additional work from user. smart_sensor_group_xxx DAGs will be included when "use_smart_sensor" configuration is True and they are handling all sensors in the airflow cluster. Users doesn't need to change any DAG definition. The smart sensor is more close to a transparent service provided by airflow core. 
  c. Configure supported sensor class by comma separated class name in "sensors_enabled" configuration. 
  d. Based on the number of sensor jobs in the airflow cluster, user can configure how many smart sensors should be enabled by setting "shards" for better handling current traffic. For our experience, each smart sensor task can easily handle 100~300 concurrently running sensor jobs. This number could be different in different clusters/use cases.

7. Simulate task instance logs that executed by smart sensor: Even if the sensor tasks were running in smart sensor process, we hold the user can easily achieve the task progress by viewing same log location.
  a. In smart sensor operator, simulate logs for all sensors handled by current task.
  b. For es handler: since logs may come from different hosts and will be messed up ordering by offset, change the returned logs from one big string into hosted logs 
 
Impact after deploying smart sensor:

System Improvement: The number of parallel sensor tasks is reduced by 85%, and the number of concurrent airflow tasks (both sensor and non-sensor tasks) is reduced more than 60% . The reduction of tasks greatly helps stabilize airflow cluster by reducing the airflow metaDB load, mostly from heartbeat operations (before deployment sensor tasks took more than 70% ratio of all running tasks). 

<img width="1812" alt="Screen Shot 2020-01-10 at 3 26 17 PM" src="https://user-images.githubusercontent.com/2688818/82954432-95acdf00-9f61-11ea-90c4-1a692f6d150d.png">

<img width="1812" alt="Screen Shot 2020-01-10 at 3 28 23 PM" src="https://user-images.githubusercontent.com/2688818/82954584-ea505a00-9f61-11ea-8d6a-68afb509c685.png">

Reducing the request to Hive metastore database because the smart sensor consolidates duplicate sensor tasks. As shown below, the gray dot-line shows the original number of requests to hive metastore database and the blue solid line shows the number of requests after we deployed the smart sensor.
<img width="1824" alt="Screen Shot 2020-01-10 at 3 30 59 PM" src="https://user-images.githubusercontent.com/2688818/82954774-44e9b600-9f62-11ea-89f6-a11f4c1a58de.png">

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
In progress

tests/sensors/test_smart_sensor_operator.py
tests/models/test_sensorinstance.py 
Made updates to scheduler job, log handler and views test.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

@KevinYang21

BasPH reviewed (dismissed):
Hi @YingboWang I agree with the idea of persisting more information of operators in the DB but disagree with creating yet another sensor type. I suggest we make sensors "smart" by default instead of creating yet another type, which would be very confusing to the Airflow users.

Will inspect the code line-by-line later.

Commit:
[AIRFLOW-3964] Consolidate and de-dup sensor tasks using smart sensor

* [AIRBNB][DI-3845] Build smart sensor operator

* [DI-3845][addendum]Support infra retry in smart sensor (#372)

* [DI-3845][addendum]Group distributed task log by host (#373)

* [DI-3845][addendum]Set the end_of_log only logs on all hosts end (#381)

Set end_of_log when logs from all know hosts have been fully loaded or
es query has lasted for 5 mins.

This is to fix the issue that sometime a second worker can pick up a running
task and exit after checking dependency. Checking only the end of log mark for
a host with later timestamp may violate the loading of real running worker.

* [DI-3845][addendum]Fix UI empty log return (#383)

* [DI-3845][addendum]Raise infra failure without retry for smart sensor (#384)

* [DI-3845][addendum]Add query infor for exception log (#387)

Apply Black formatting

Apply Black formatting

turbaszek commented on airflow/models/sensorinstance.py:
```diff
@@ -0,0 +1,169 @@
+            elif hasattr(builtin_sensors, class_name):
+                stem = 'airflow.sensors.'
+            elif hasattr(contrib_operators, class_name):
+                stem = 'airflow.contrib.operators.'
```
Should we get rid of `contrib` here?

YingboWang replied:
Good catch. Removed the contrib.

turbaszek commented on airflow/models/taskinstance.py:
```diff
@@ -1183,6 +1188,24 @@ def signal_handler(signum, frame):  # pylint: disable=unused-argument
+                self.log.warning("Failed to register in sensor service."
+                                 "Continue to run task in non smart sensor mode.")
+                self.log.exception(e, exc_info=True)
+
```
Should it be possible to simplify this part?
```python
registered = False
try:
    registered = task_copy.register_in_sensor_service(self, context)
except Exception as e:
    self.log.warning("Failed to register in sensor service."
                                  "Continue to run task in non smart sensor mode.")
    self.log.exception(e, exc_info=True)

if registered:
    self.log.info('Submitting %s to sensor service', self)
    self.state = State.SENSING
    self.start_date = timezone.utcnow()
    session.merge(self)
    session.commit()
    # Raise exception for sensing state
    raise AirflowSmartSensorException("Task successfully registered in smart 
```
WDYT? The last part can be encapsulated into a helper method

YingboWang replied:
Good point. Changed.

turbaszek commented on airflow/providers/elasticsearch/log/es_task_handler.py:
```diff
@@ -35,6 +36,9 @@
 from airflow.utils.log.logging_mixin import LoggingMixin
 
+# Elasticsearch hosted log type
+EsLogMsgType = List[Tuple[str, str]]
```
Are those related changes?

YingboWang replied:
Since the execution of one task instance will be processed in two steps, the logs generated from these two steps might come from different host. In es_task_handler, the logs are fetched by offset and it could mess up logs from two hosts. The change in here is to change the returning logs from string to a list of hosted log string so that users can have a readable log url.

turbaszek commented on docs/smart-sensor.rst:
```diff
@@ -0,0 +1,86 @@
+============
+
+The smart sensor is a service which greatly reduces airflow’s infrastructure cost by consolidating
+some of the airflow long running light weight tasks.DAG Serialization and DB Persistence.
```
```suggestion
some of the airflow long running light weight tasks. DAG Serialization and DB Persistence.
```

turbaszek commented on docs/smart-sensor.rst:
```diff
@@ -0,0 +1,86 @@
+store poke context at sensor_instance table and then exits with a ‘sensing’ state.
+
+When the smart sensor mode is enabled, a special set of builtin smart sensor DAGs
+(named smart_sensor_group_shard_xxx) is created by the system; These DAGs contain SmartSensorOperator
```
```suggestion
(named smart_sensor_group_shard_xxx) is created by the system; These DAGs contain ``SmartSensorOperator``
```

turbaszek commented on docs/smart-sensor.rst:
```diff
@@ -0,0 +1,86 @@
+Support new operators in the smart sensor service
+-------------------------------------------------
+
+*   Define ``poke_context_fields`` as class attribute in the operator. ``poke_context_fields``
```
```suggestion
*   Define ``poke_context_fields`` as class attribute in the sensor. ``poke_context_fields``
```
Right?

YingboWang replied:
Fixed.

kaxil commented on airflow/config_templates/config.yml:
```diff
@@ -2331,3 +2331,35 @@
+  options:
+    - name: use_smart_sensor
+      description: |
+        When use smart sensor, redirect multiple qualified sensor tasks to smart sensor task
```
```suggestion
        When `use_smart_sensor` is True, Airflow redirects multiple qualified sensor tasks to smart sensor task
```

YingboWang replied:
Followed the suggestion. Thanks

kaxil commented on airflow/migrations/versions/e38be357a868_update_schema_for_smart_sensor.py:
```diff
@@ -0,0 +1,70 @@
+        sa.Column('poke_context', sa.Text(), nullable=False),
+        sa.Column('execution_context', sa.Text(), nullable=True),
+        sa.Column('created_at', sa.TIMESTAMP(timezone=True), default=func.now(), nullable=False),
+        sa.Column('updated_at', sa.TIMESTAMP(timezone=True), default=func.now(), nullable=False),
```
Any reason for having the defaults here? (Just curious)

kaxil replied:
Also I think the Timestamp needs to have precision for MySQL. Similar to https://github.com/apache/airflow/commit/9e6b5abea08eb35c48839e33a4518f292ef7564e#diff-b89252a0e7858101fe0bbd8bf89faf1aR40-R49

YingboWang replied:
Should I create table with different types based on the` conn.dialect` here? Or do an `alter table` after?

kaxil replied:
Yes similar to https://github.com/apache/airflow/blob/master/airflow/migrations/versions/0a2a5b66e19d_add_task_reschedule_table.py

kaxil commented on airflow/models/sensorinstance.py:
```diff
@@ -0,0 +1,166 @@
+    def get_classpath(obj):
+        """
+        Get the object dotted class path. Used for getting operator classpath
+        :param obj:
```
Needs docstrings

YingboWang replied:
Thanks. Blank space added.

kaxil commented on airflow/models/sensorinstance.py:
```diff
@@ -0,0 +1,166 @@
+    def get_classpath(obj):
+        """
+        Get the object dotted class path. Used for getting operator classpath
+        :param obj:
```
Needs a blank space between description and params for Sphinx rendering

```suggestion

        :param obj:
```

kaxil commented on airflow/models/sensorinstance.py:
```diff
@@ -0,0 +1,166 @@
+        module_name, class_name = obj.__module__, obj.__class__.__name__
+
+        if '.' not in module_name:
+            # Fix if module name was broken by airflow importer
```
When can this happen?

YingboWang replied:
Good catch. This is to fix the old airflow importer issue which I believe is deprecated now. I will remove this block.

kaxil commented on airflow/models/sensorinstance.py:
```diff
@@ -0,0 +1,166 @@
+        :param ti: The task instance for the sensor to be registered.
+        :param poke_context: Context used for sensor poke function.
+        :param execution_context: Context used for execute sensor such as timeout
+        setting and email configuration.
```
```suggestion
    def register(cls, ti, poke_context, execution_context, session=None):
        """
        Register task instance ti for a sensor in sensor_instance table. Persist the
        context used for a sensor and set the sensor_instance table state to sensing.
        
        :param ti: The task instance for the sensor to be registered.
        :param poke_context: Context used for sensor poke function.
        :param execution_context: Context used for execute sensor such as timeout
            setting and email configuration.
```

kaxil commented on airflow/models/baseoperator.py:
```diff
@@ -1338,6 +1338,13 @@ def get_serialized_fields(cls):
+    def is_smart_sensor_compatible(self):
+        """
+        Return if this operator can use smart service. Default False.
+        :return:
```
```suggestion
```

kaxil commented on airflow/models/taskinstance.py:
```diff
@@ -280,7 +281,8 @@ def try_number(self):
         # This is designed so that task logs end up in the right file.
-        if self.state == State.RUNNING:
+        # TODO: whether we need sensing here or not (in sensor and task_instance state machine)
+        if self.state in State.running():
```
Just to confirm here, the state of SENSING just means the "Sensor DAG" has this task, or does it mean this TI is currently being Poked ?

YingboWang replied:
"sensing" state means this TI is currently being poked by smart sensor.

kaxil commented on airflow/providers/apache/hive/sensors/named_hive_partition.py:
```diff
@@ -104,3 +105,9 @@ def poke(self, context: Dict[str, Any]) -> bool:
+    def is_smart_sensor_compatible(self):
+        result = not self.soft_fail and not self.hook and \
+            len(self.partition_names) <= 30 and \
+            super(NamedHivePartitionSensor, self).is_smart_sensor_compatible()
```
```suggestion
            super().is_smart_sensor_compatible()
```

kaxil commented on airflow/sensors/base_sensor_operator.py:
```diff
@@ -66,6 +68,8 @@ class BaseSensorOperator(BaseOperator, SkipMixin):
     """
     ui_color = '#e6f1f2'  # type: str
     valid_modes = ['poke', 'reschedule']  # type: Iterable[str]
+    execution_fields = ('poke_interval', 'retries', 'execution_timeout', 'timeout',
```
Can we add a comment on what should go in `execution_fields`

kaxil commented on airflow/sensors/base_sensor_operator.py:
```diff
@@ -106,6 +113,62 @@ def poke(self, context: Dict) -> bool:
+        """
+        Register ti in smart sensor service
+        :param ti:
+        :param context:
```
Needs to be filled in

kaxil commented on airflow/sensors/base_sensor_operator.py:
```diff
@@ -106,6 +113,62 @@ def poke(self, context: Dict) -> bool:
+        Return a dictionary with all attributes in poke_context_fields. The
+        poke_context with operator class can be used to identify a unique
+        sensor job.
+        :param context:
```
Here too

kaxil commented on airflow/sensors/base_sensor_operator.py:
```diff
@@ -106,6 +113,62 @@ def poke(self, context: Dict) -> bool:
+        Return a dictionary with all attributes in execution_fields. The
+        execution_context include execution requirement for each sensor task
+        such as timeout setup, email_alert setup.
+        :param context:
```
Same here

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+    """
+    This class stores a sensor work with decoded context value. It is only used
+    inside of smart sensor.
+    """
```
Can we add docstrings for params

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+
+    def _get_sensor_logger(self, ti):
+        # TODO: should be somewhere else, but not this file, has to use LOG_ID_TEMPLATE from es
+        # but how about other log file handler?
```
Did you'll decide anything on this TODO?

YingboWang replied:
Removed.

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+        ti.raw = False  # Otherwise set_context will fail
+        log_id = "-".join([ti.dag_id,
+                           ti.task_id,
+                           ti.execution_date.strftime("%Y_%m_%dT%H_%M_%S_%f"),
```
Wil this fail if users have configured a different log-template-id

YingboWang replied:
Should not fail. The log_id is used as an internal key for fetching these in memory log handler only in smart sensor service. The output path was configure in the `set_context` function which should always be consistent with other tasks.

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+class CachedPokeWork:
+    """
+    Wrapper class for the poke work inside smart sensor. It saves
+    the sensor_task used to poke and recent poke result state
```
needs docstrings here

kaxil replied:
i.e. info regarding args

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+    :type poke_interval: int
+    :param timeout: Time, in seconds before the task times out and fails.
+    :type timeout: int
+    :type mode: str
```
Some of them are missing:

- shard_min
- shard_max
- poke_exception_cache_ttl
- poke_timeout
- poke_exception_to_fail_task_threshold

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+            .filter(SI.shardcode < self.shard_max,
+                    SI.shardcode >= self.shard_min)
+        tis = query.all()
+        session.commit()
```
Do we need this? We are simply querying in this function so I think we can safely remove this

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+
+        count_marked = 0
+        try:
+            query_result = session.query(TI, SI)\
+                .join(TI, and_(TI.dag_id == SI.dag_id,
+                               TI.task_id == SI.task_id,
+                               TI.execution_date == SI.execution_date)) \
+                .filter(SI.state == State.SENSING) \
+                .filter(SI.hashcode == poke_hash) \
+                .filter(SI.operator == operator) \
+                .with_for_update().all()
```
I am concerned about any join that involves TaskInstance as that table can contain large number of records.

Did you find any issues with this with your Prod cluster?

kaxil replied:
How about if we filter first with `State.SENSING` to decrease the set we do joins, WDYT?

YingboWang replied:
Thanks Kaxil for raising this. Yes the query performance is really important. I think this query should be safe at least in mysql engine. 

We have the index on the hashcode in sensor_instace table which can largely reduce the time of this query (and also another query for load sensor tasks).

In our experience of using smart sensor, this query did not cause any issue and the max duplicate jobs in our db is around 500 at the same time. We all know that our scale of airflow cluster is much larger than most use cases so I assume if this query is safe in our cluster it seems not likely to cause a performance issue in other system. 

Also, before creating this PR, we have tested with `explain` the queries with different order in our mysql db. For this query the results are the same. DB engine do the optimization and putting the `filter` before or after the `join` in sqlalchemy does not really matter.

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+                    mark_state(ti=ti, sensor_instance=sensor_instance)
+                    count_marked += 1
+                else:
+                    # ti.state != State.SENSING
```
Can this be removed?

YingboWang replied:
Task instance states could be changed by user activities such as marking dagrun and ti states from UI. The `sensor_instance` table will respect `task_instance` state changing that happened during it's poking phase. So if a user mark a ti state, it will not be overwritten by smart sensor.

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+            self.log.exception("Exception _mark_multi_state in smart sensor for hashcode %s",
+                               str(poke_hash))
+            self.log.exception(e, exc_info=True)
+        self.log.info("Marked %s tasks out of %s to state %s", count_marked, len(query_result), state)
```
https://docs.python.org/3/library/logging.html#logging.Logger.exception
>exception(msg, *args, **kwargs)¶
>Logs a message with level ERROR on this logger. The arguments are interpreted as for debug(). Exception info is added to the logging message. This method should only be called from an exception handler.

So I think the 2nd `self.log.exception` can be removed

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+        :param ti_key:
+        :param state:
+        :param session:
+        :return:
```
Needs info

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+        """
+        Change single task state for sensor task. For final state, set the end_date.
+        Since smart sensor take care all retries in one process. Failed sensor tasks
+        logically experienced all retries and the try_nubmer should be set to max_tries.
```
```suggestion
        logically experienced all retries and the try_number should be set to max_tries.
```

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,730 @@
+        started_at = timezone.utcnow()
+
+        self.hostname = get_hostname()
+        while True:
```
When they are no pokes needed, does the smart sensor DAG keeps running or is it marked complete

YingboWang replied:
Smart sensor DAGs keeps running in smart sensor mode. One task will be turned off by the execution_timeout after running for 24 hours and there will always a new dagrun kicked off to take over the poking job.

kaxil commented on airflow/smart_sensor_dags/smart_sensor_group.py:
```diff
@@ -0,0 +1,63 @@
+
+args = {
+    'owner': 'airflow',
+    'start_date': days_ago(2),
```
```suggestion
```

kaxil replied:
`start_date` should only be used on task level if a user likes to have a different `start_date` for that particular task. Adding `start_date` in `default_args` basically means all tasks have same start_date, so it is better to define them at DAG level instead. Check https://github.com/apache/airflow/pull/9982 for more details

kaxil commented on airflow/smart_sensor_dags/smart_sensor_group.py:
```diff
@@ -0,0 +1,63 @@
+    dag = DAG(
+        dag_id=dag_id,
+        default_args=args,
+        schedule_interval=timedelta(days=1),
```
```suggestion
        schedule_interval=timedelta(days=1),
        start_date=days_ago(2),
```

YingboWang replied:
The same start_date was defined in default_args.

kaxil replied:
`start_date` should only be used on task level if a user likes to have a different `start_date` for that particular task. Adding `start_date` in `default_args` basically means all tasks have same start_date, so it is better to define them at DAG level instead. Check https://github.com/apache/airflow/pull/9982 for more details

kaxil commented on airflow/utils/log/file_processor_handler.py:
```diff
@@ -83,7 +83,14 @@ def close(self):
     def _render_filename(self, filename):
-        filename = os.path.relpath(filename, self.dag_dir)
+        # Fix for airflow native dag access right problem. If the filename is in airflow
+        # source, mock as the dag is under dag_dir/native_dags
```
Can you expand on this a little bit -- I am not sure I understand

kaxil commented on airflow/utils/log/file_task_handler.py:
```diff
@@ -71,15 +71,25 @@ def close(self):
                                              execution_date=ti.execution_date.isoformat(),
                                              try_number=try_number)
 
+    def _read_grouped_logs(self):
+        return False
```
Is it supposed to return `False` always ?

YingboWang replied:
For es_task_handler it will return True.

kaxil replied:
got it

kaxil commented on docs/smart-sensor.rst:
```diff
@@ -0,0 +1,86 @@
+Smart Sensor
+============
+
+The smart sensor is a service which greatly reduces airflow’s infrastructure cost by consolidating
```
```suggestion
The smart sensor is a service (run by a builtin DAG) which greatly reduces airflow’s infrastructure cost by consolidating
```

YingboWang replied:
Updated.

kaxil commented on docs/smart-sensor.rst:
```diff
@@ -0,0 +1,86 @@
+
+.. image:: img/smart_sensor_architecture.png
+
+Instead of using one process for each task, the main idea of the smart sensor service to improve the
```
```suggestion
Instead of using one process for each task, the main idea of the smart sensor service is to improve the
```

kaxil reviewed (commented):
Great work @YingboWang 🎉 . Sorry it took so long for the review

kaxil reviewed (commented):
There are some other comments, that might be hidden, you could check them at https://github.com/apache/airflow/pull/5499/files

kaxil commented on airflow/sensors/smart_sensor_operator.py:
```diff
@@ -0,0 +1,764 @@
+                if cached_exception.fail_current_run:
+                    count_exception_failures += 1
+
+            Stats.gauge("smart_sensor_operator.poked_tasks", count_poke)
+            Stats.gauge("smart_sensor_operator.poked_success", count_poke_success)
+            Stats.gauge("smart_sensor_operator.poked_exception", count_poke_exception)
+            Stats.gauge("smart_sensor_operator.exception_failures", count_exception_failures)
+            Stats.gauge("smart_sensor_operator.infra_failures", count_infra_failure)
```
Can you please add these metrics to https://github.com/apache/airflow/blob/master/docs/metrics.rst

YingboWang replied:
Thanks for reminding. Added metrics to the doc.

kaxil commented on airflow/smart_sensor_dags/smart_sensor_group.py:
```diff
@@ -0,0 +1,64 @@
+        shard_max=shard_max,
+        poke_timeout=10,
+        smart_sensor_timeout=timedelta(hours=24).total_seconds(),
+        start_date=days_ago(2),
```
Can we move the `start_date` from the Task to DAG instead please?

YingboWang replied:
Done

kaxil reviewed (approved):
If anyone has any more suggestions or want to request changes, let's do it in a follow-up PR. 

Thanks alot @YingboWang and apologies for the long wait.
